### PR TITLE
Causal Imitation Learning Algorithm

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Implementation of the causal imitation learning algorithm proposed in
[Swamy et al. "Causal Imitation Learning under Temporally Correlated Noise"  ICML2022](https://proceedings.mlr.press/v162/swamy22a/swamy22a.pdf).

### CODE REVIEW

File: alf/utils/dist_utils.py [hunk 1]
```diff
@@ -1158,6 +1158,60 @@ def get_mode(dist):
...
+        The mode of the distribution. If ``dist`` is a transformed distribution,
+        the result is calculated by transforming the mode of its base
+        distribution and may not be the actual mode for ``dist``.
+    Raises:
+        NotImplementedError: if dist or its base distribution is not
+            ``td.Categorical``, ``td.Normal``, ``td.Independent`` or
+            ``td.TransformedDistribution``.
+    """
+    if isinstance(dist, td.categorical.Categorical):
+        mode = torch.argmax(dist.logits, -1)
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: This does not support backpropagation

File: alf/utils/dist_utils.py [hunk 2]
```diff
@@ -1158,6 +1158,60 @@ def get_mode(dist):
...
+    elif isinstance(
+            dist,
+        (OneHotCategoricalStraightThrough, OneHotCategoricalGumbelSoftmax)):
+        # Our version of one-hot st supports mode with grad
+        mode = dist.mode
+    elif isinstance(
+            dist, (td.OneHotCategorical, td.OneHotCategoricalStraightThrough)):
+        mode = torch.nn.functional.one_hot(
+            torch.argmax(dist.logits, -1), num_classes=dist.logits.shape[-1])
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: cannot backprop for this

File: alf/utils/dist_utils.py [hunk 3]
```diff
@@ -1158,6 +1158,60 @@ def get_mode(dist):
...
+        mode = base_mode
+        for transform in dist.transforms:
+            mode = transform(mode)
+    elif isinstance(dist, Beta):
+        alpha = dist.concentration1
+        beta = dist.concentration0
+        mode = torch.where((alpha > 1) & (beta > 1),
+                           (alpha - 1) / (alpha + beta - 2),
+                           torch.where(alpha < beta, torch.zeros(()),
+                                       torch.ones(())))
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: implement this as a property function of Beta

File: alf/algorithms/causal_bc_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            inputs.observation, state=state.actor)
+        action = dist_utils.epsilon_greedy_sample(action_dist,
+                                                  self._epsilon_greedy)
+
+        return AlgStep(output=action, state=BcState(actor=new_state))
+
+    def residuIL_loss(self, targets, predictions, pred_residuals):
+
+        # train policy (detach discriminator)
+        policy_loss = (2 * (targets - predictions) * pred_residuals.detach()
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py]: "targets-predictions" is used multiple times. Should precompute it.

File: alf/algorithms/causal_bc_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                           torch.square(targets - predictions)).mean(-1)
+
+        # train discriminator (detach policy)
+        discriminator_loss = -(
+            2 * (targets - predictions).detach() * pred_residuals -
+            pred_residuals * pred_residuals).mean(-1)
+
+        discriminator_loss = (
+            discriminator_loss +
+            self._f_norm_penalty_weight * torch.linalg.norm(pred_residuals))
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py:176]: I cannot find the description of norm penalty in the paper.

File: alf/algorithms/causal_bc_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        action = dist_utils.epsilon_greedy_sample(action_dist,
+                                                  self._epsilon_greedy)
+
+        return AlgStep(output=action, state=BcState(actor=new_state))
+
+    def residuIL_loss(self, targets, predictions, pred_residuals):
+
+        # train policy (detach discriminator)
+        policy_loss = (2 * (targets - predictions) * pred_residuals.detach()
+                       ).mean(-1) + self._bc_regulatization_weight * (
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py]: I cannot find the description of bc regularization weight in the paper.

File: alf/algorithms/causal_bc_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        Swamy et al. Causal Imitation Learning under Temporally Correlated Noise,
+        ICML 2022
+    """
+
+    def __init__(self,
+                 observation_spec,
+                 action_spec: BoundedTensorSpec,
+                 reward_spec=TensorSpec(()),
+                 actor_network_cls=ActorNetwork,
+                 discriminator_network_cls=EncodingNetwork,
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py:50]: Need to comment this argument.

File: alf/algorithms/causal_bc_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        info = BcInfo(
+            actor=predictions,
+            discriminator=pred_residuals,
+            target=rollout_info.action)
+        return AlgStep(
+            rollout_info.action, state=BcState(actor=new_state), info=info)
+
+    def calc_loss_offline(self, info, pre_train=False):
+
+        #[T, B, 1]
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py]: "1" should be the dimension of action?

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/utils/dist_utils.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: Removed

File: alf/utils/dist_utils.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: Removed

File: alf/utils/dist_utils.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: Good suggestion. Change as Beta's property function.

File: alf/algorithms/causal_bc_algorithm.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/algorithms/causal_bc_algorithm.py]: Changed.

File: alf/algorithms/causal_bc_algorithm.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/algorithms/causal_bc_algorithm.py:176]: This penalty is mentioned in their Appendix Table 4 and 5.

File: alf/algorithms/causal_bc_algorithm.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/algorithms/causal_bc_algorithm.py]: This term is also mentioned in their Appendix Table 4 and 5.

File: alf/algorithms/causal_bc_algorithm.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/algorithms/causal_bc_algorithm.py:50]: Added

File: alf/algorithms/causal_bc_algorithm.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/algorithms/causal_bc_algorithm.py]: Yes, should be action dimension. Updated.

File: alf/algorithms/causal_bc_algorithm.py [hunk 9]
```diff
@@ -71,6 +71,10 @@ def __init__(self,
             actor_network_cls (Callable): is used to construct the actor network.
                 The constructed actor network is a determinstic network and
                 will be used to generate continuous actions.
+            discriminator_network_cls (Callable): is used to construct the
+                discriminator network. The discrimonator is trained in a way
+                that is adversarial to the training of the policy, ot help with
```
Comment (emailweixu) [inline alf/algorithms/causal_bc_algorithm.py]: ot => it
comment its expected input and output (not activated)

File: alf/utils/dist_utils.py [hunk 10]
```diff
@@ -1158,6 +1161,42 @@ def get_mode(dist):
...
+    Raises:
+        NotImplementedError: if dist or its base distribution is not
+            ``td.Normal``, ``StableCauchy``, ``Beta``, ``TruncatedDistribution``,
+            ``td.Independent`` or ``td.TransformedDistribution``.
+    """
+    if isinstance(dist, td.normal.Normal):
+        mode = dist.mean
+    elif isinstance(dist, StableCauchy):
+        mode = dist.loc
+    elif isinstance(dist, Beta) or isinstance(dist, TruncatedDistribution):
```
Comment (emailweixu) [inline alf/utils/dist_utils.py:1184]: Can be simplified as `isinstance(dist, (Beta, TruncatedDistribution))`

**Review by emailweixu — APPROVED** [review]

**Pull request merged**